core/src/compile/state.rs

```diff
@@ -0,0 +1,578 @@
+impl<'a, G: Generator> Default for StructuredState<'a, G> {
+    fn default() -> Self {
+        Self {
+            children: BTreeMap::new(),
+            ordering: Vec::new(),
+        }
+    }
+}
+
+pub mod structured_state {
+    use std::collections::BTreeMap;
+
+    use synth_gen::prelude::Generator;
+
+    use super::{CompilerState, StructuredState};
+
+    pub struct IntoIter<'a, G: Generator> {
+        children: BTreeMap<String, CompilerState<'a, G>>,
+        iter: std::vec::IntoIter<String>,
+    }
+
+    impl<'a, G: Generator> From<StructuredState<'a, G>> for IntoIter<'a, G> {
+        #[inline]
+        fn from(v: StructuredState<'a, G>) -> Self {
+            Self {
+                children: v.children,
+                iter: v.ordering.into_iter(),
+            }
+        }
+    }
+
+    impl<'a, G: Generator> Iterator for IntoIter<'a, G> {
+        type Item = (String, CompilerState<'a, G>);
+
+        #[inline]
+        fn next(&mut self) -> Option<Self::Item> {
+            self.iter
+                .next()
+                .and_then(|name| self.children.remove(&name).map(|state| (name, state)))
+        }
+    }
+}
+
+pub enum Source<'a> {
+    Namespace(&'a Namespace),
+    Content(&'a Content),
+}
+
+impl<'a> std::fmt::Display for Source<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::Namespace(_) => write!(f, "namespace"),
+            Self::Content(_) => write!(f, "content"),
+        }
+    }
+}
+
+impl<'a> Source<'a> {
+    #[inline]
+    pub fn as_namespace(&self) -> Result<&'a Namespace> {
+        match self {
+            Self::Namespace(namespace) => Ok(namespace),
+            other => Err(failed!(
+                target: Release,
+                "source node type mismatch: expected namespace, found {}",
+                other
+            )),
+        }
+    }
+
+    #[inline]
+    pub fn as_content(&self) -> Result<&'a Content> {
+        match self {
+            Self::Content(content) => Ok(content),
+            other => Err(failed!(
+                target: Release,
+                "source node type mismatch: expected content, found {}",
+                other
+            )),
+        }
+    }
+}
+
+pub enum Artifact<G, Y, R> {
+    Just(G),
+    Link(Link<G, Y, R>),
+}
+
+impl<G, Y, R> Artifact<G, Y, R>
+where
+    G: Generator<Yield = Y, Return = R> + FromLink<Yield = Y, Return = R>
+{
+    pub(super) fn pack(self) -> G {
+        match self {
+            Self::Just(g) => g,
+            Self::Link(link) => G::from_link(link),
+        }
+    }
+
+    pub(super) fn just(inner: G) -> Self {
+        Self::Just(inner)
+    }
+
+    pub(super) fn from_view(view: TapeView<Y, R>) -> Self {
+        Self::Link(Link::View(view))
+    }
+
+    pub(super) fn from_recorder(recorder: Recorder<G, Y, R>) -> Self {
+        Self::Link(Link::Recorder(recorder))
+    }
+}
+
+pub type GenArtifact<G: Generator> = Artifact<G, G::Yield, G::Return>;
+
```

Comment:
I would also lightly document the strate transition here and the 'meaning' of each state :)